File: Data Structure/Linked List/Medium/LET_no147_Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        if (!head || !(head->next))
            return head;
        
        ListNode* dummy = new ListNode(0, head);
        ListNode *prev = dummy;
        
        while (head) {
            if (head->next && head->next->val < head->val) {
                while (prev->next && prev->next->val < head->next->val)
                    prev = prev->next;
                
                ListNode* temp = prev->next;
                prev->next = head->next;
                head->next = head->next->next;
                prev->next->next = temp;
                prev = dummy;
            }
            else
                head = head->next;
        }
        
        return dummy->next;
    }
};

File: Data Structure/Linked List/Medium/LET_no147_Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        dummy = ListNode(-1)
        while head:
            cur = dummy
            nxt = head.next

            while cur.next and cur.next.val < head.val:
                cur = cur.next

            head.next = cur.next
            cur.next = head
            head = nxt
        
        return dummy.next